tsl/src/compression/api.c

```diff
@@ -892,3 +976,655 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+				{
+					segment_info = segment_info_new(TupleDescAttr(slot->tts_tupleDescriptor, col));
+					current_segment[i]->decompressed_chunk_offset =
+						decompressor.per_compressed_cols[col].decompressed_column_offset;
+					/* also need to call segment_info_update here to update the val part */
+					segment_info_update(segment_info, val, is_null);
+					current_segment[i]->segment_info = segment_info;
+					i++;
+				}
+			}
+		}
+		/* we have a segment already, so compare those */
+		changed_segment = decompress_segment_changed_group(current_segment,
+														   slot,
+														   &decompressor.per_compressed_cols);
+		if (!changed_segment)
+		{
+			i = 0;
+			bool should_free;
+
+			compressed_tuple = ExecFetchSlotHeapTuple(slot, false, &should_free);
+
+			heap_deform_tuple(compressed_tuple,
+							  compressed_rel_tupdesc,
+							  decompressor.compressed_datums,
+							  decompressor.compressed_is_nulls);
+
+			row_decompressor_decompress_row(&decompressor, segment_tuplesortstate);
+
+			simple_table_tuple_delete(compressed_chunk_rel, &(slot->tts_tid), snapshot);
+
+			if (should_free)
+				heap_freetuple(compressed_tuple);
+			ExecClearTuple(slot);
+		}
+		else if (changed_segment)
+		{
+			fetch_matching_uncompressed_chunk_into_tuplestore(segment_tuplesortstate,
+															  nsegmentby_cols,
+															  uncompressed_chunk_rel,
+															  current_segment);
+
+			tuplesort_performsort(segment_tuplesortstate);
+
+			recompress_segment(segment_tuplesortstate, uncompressed_chunk_rel, &row_compressor);
+
+			tuplesort_end(segment_tuplesortstate); // now any pointers returned will be garbage
+
+			/* make changes visible */
+			CommandCounterIncrement();
+
+			decompress_segment_update_current_segment(current_segment,
+													  slot, /*slot from compressed chunk*/
+													  decompressor.per_compressed_cols);
+			/* reinit tuplesort and add the first tuple of the new segment to it */
+			segment_tuplesortstate = tuplesort_begin_heap(uncompressed_rel_tupdesc,
+														  n_keys,
+														  sort_keys,
+														  sort_operators,
+														  sort_collations,
+														  nulls_first,
+														  maintenance_work_mem,
+														  NULL,
+														  false);
+
+			bool should_free;
+			compressed_tuple = ExecFetchSlotHeapTuple(slot, false, &should_free);
+
+			heap_deform_tuple(compressed_tuple,
+							  compressed_rel_tupdesc,
+							  decompressor.compressed_datums,
+							  decompressor.compressed_is_nulls);
+
+			row_decompressor_decompress_row(&decompressor, segment_tuplesortstate);
+
+			simple_table_tuple_delete(compressed_chunk_rel, &(slot->tts_tid), snapshot);
+			changed_segment = false; // because this is the first tuple of the new segment
+			/* make changes visible */
+			CommandCounterIncrement();
+
+			if (should_free)
+				heap_freetuple(compressed_tuple);
+			ExecClearTuple(slot);
```

Comment:
I am not sure if I follow the code flow here. We have a while loop over `index_getnext_slot`. The slot is cleared in the `if` branch (Line 1505) and here in the `if else` branch. Shouldn't it be cleared in every iteration of the while loop?

Reply:
Yes you are right, thank you for pointing this out. Fixed.